Examples/WindowsFormsApplication/Nfield.cs

```diff
@@ -0,0 +1,138 @@
+﻿using Microsoft.Identity.Client;
+using Microsoft.Win32;
+using Nfield.Infrastructure;
+using Nfield.Models;
+using Nfield.Services;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace WindowsFormsApplication
+{
+    public class Nfield
+    {
+        private IPublicClientApplication ClientApp { get; init; }
+        private string[] Scopes { get; } = new[] { CreateScope("user_impersonation"), CreateScope("Survey.Read") };
+        public string DomainName { get; }
+
+        private INfieldConnectionV2 NfieldConnection { get; set; }
+
+        public Nfield(string domainName)
+        {
+            ClientApp = CreateClientApp();
+            DomainName = domainName;
+
+            NfieldConnection = NfieldConnectionFactory.Create(ApplicationConfiguration.Current.NfieldApiUri);
+            NfieldConnection.RegisterTokenProvider(domainName, ProvideTokenAsync);
+        }
+
+        private IPublicClientApplication CreateClientApp()
+        {
+            var client = PublicClientApplicationBuilder.Create(ApplicationConfiguration.Current.ClientId)
+                .WithRedirectUri("http://localhost")
+                .WithAuthority(AzureCloudInstance.AzurePublic, ApplicationConfiguration.Current.Tenant)
+                .Build();
+
+            client.UserTokenCache.SetBeforeAccess(BeforeAccessNotification);
+            client.UserTokenCache.SetAfterAccess(AfterAccessNotification);
+
+            return client;
+        }
+
+        private async Task<string> ProvideTokenAsync()
+        {
+            var result = await AuthenticateAsync((status) => { });
+            return result.AccessToken;
+        }
+
+        private static string CreateScope(string scope) => $"{ApplicationConfiguration.Current.NfieldApiApplicationId}/{scope}";
+
+        internal async Task LogoutAsync()
+        {
+            var accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            while (accounts.Any())
+            {
+                await ClientApp.RemoveAsync(accounts.First());
+                accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            }
+        }
+
+        private byte[] RestoreTokenCacheFromRegistry()
+        {
+            using (RegistryKey key = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\NIPO Samples"))
+            {
+                var value = key.GetValue("TokenCache");
```

Comment:
Is it common practice to store token cache in  registry? shouldnt we use the special windows sdk for save consolidated app storage (since tokens are critical/security sensitive). I think there are some secure data api's or something in windows to deal better with this.

Reply:
Registry is quite common for client apps, I now added data protection scoped to the current user to it, so only that user can decrypt the token data.